### PR TITLE
MPS Implementation p5: InnerProduct

### PR DESCRIPTION
**Pull request by MichaelBroughton**
Adds the InnerProduct functionality to `mps_statespace.h`

### CODE REVIEW

File: lib/mps_statespace.h [hunk 1]
```diff
@@ -155,6 +155,63 @@ class MPSStateSpace {
     }
   }
 
+  // Computes <state1 | state2 > for two equal sized MPS.
+  // Requires: state1.bond_dim() == state2.bond_dim() &&
+  //           state1.num_qubits() == state2.num_qubits()
+  static std::complex<fp_type> InnerProduct(MPS& state1, MPS& state2) {
```
Comment (sergeisakov) [inline lib/mps_statespace.h:161]: Perhaps add `const`: `MPS& state1` -> `const MPS& state1` and `MPS& state2` -> `const MPS& state2`?

File: lib/mps_statespace.h [hunk 2]
```diff
@@ -155,6 +155,63 @@ class MPSStateSpace {
...
+    const auto nq = state1.num_qubits();
+    const auto bd = state1.bond_dim();
+    const auto end = Size(state1);
+    auto offset = 0;
+    fp_type* state1_raw = state1.get();
+    fp_type* state2_raw = state2.get();
+
+    // Contract leftmost blocks together, store result in state1 scratch.
+    ConstMatrixMap top((Complex*)state2_raw, 2, bd);
```
Comment (sergeisakov) [inline lib/mps_statespace.h:170]: Add a space between `(Complex*)` and `state` in these two lines?

File: lib/mps_statespace.h [hunk 3]
```diff
@@ -155,6 +155,63 @@ class MPSStateSpace {
     }
   }
 
+  // Computes <state1 | state2 > for two equal sized MPS.
+  // Requires: state1.bond_dim() == state2.bond_dim() &&
+  //           state1.num_qubits() == state2.num_qubits()
+  static std::complex<fp_type> InnerProduct(MPS& state1, MPS& state2) {
+    const auto nq = state1.num_qubits();
+    const auto bd = state1.bond_dim();
```
Comment (sergeisakov) [inline lib/mps_statespace.h:163]: `bond_dim` instead of `bd`?

File: lib/mps_statespace.h [hunk 4]
```diff
@@ -155,6 +155,63 @@ class MPSStateSpace {
     }
   }
 
+  // Computes <state1 | state2 > for two equal sized MPS.
+  // Requires: state1.bond_dim() == state2.bond_dim() &&
+  //           state1.num_qubits() == state2.num_qubits()
+  static std::complex<fp_type> InnerProduct(MPS& state1, MPS& state2) {
+    const auto nq = state1.num_qubits();
```
Comment (sergeisakov) [inline lib/mps_statespace.h:162]: `num_qubits` instead of `nq`?

File: lib/mps_statespace.h [hunk 5]
```diff
@@ -155,6 +155,63 @@ class MPSStateSpace {
...
+    // Contract leftmost blocks together, store result in state1 scratch.
+    ConstMatrixMap top((Complex*)state2_raw, 2, bd);
+    ConstMatrixMap bot((Complex*)state1_raw, 2, bd);
+    MatrixMap partial_contract((Complex*)(state1_raw + end), bd, bd);
+    MatrixMap partial_contract2((Complex*)(state1_raw + end + 2 * bd * bd), bd,
+                                2 * bd);
+    partial_contract.noalias() = top.adjoint() * bot;
+
+    // Contract all internal blocks together.
+    for (unsigned i = 1; i < nq - 1; i++) {
```
Comment (sergeisakov) [inline lib/mps_statespace.h:178]: Please use `++i`.

**Review by sergeisakov — CHANGES_REQUESTED** [review]

File: lib/mps_statespace.h (hunk 1 above)
Comment (MichaelBroughton) [reply lib/mps_statespace.h:166]: In the code I do write to the scratch spaces in both `state` variables so I can't make them const. Compiler complains about casting away constness if I do.

File: lib/mps_statespace.h (hunk 2 above)
Comment (MichaelBroughton) [reply lib/mps_statespace.h]: Added.

File: lib/mps_statespace.h (hunk 3 above)
Comment (MichaelBroughton) [reply lib/mps_statespace.h]: Done.

File: lib/mps_statespace.h (hunk 4 above)
Comment (MichaelBroughton) [reply lib/mps_statespace.h]: Done.

File: lib/mps_statespace.h (hunk 5 above)
Comment (MichaelBroughton) [reply lib/mps_statespace.h]: Done.

**Review by sergeisakov — APPROVED** [review]
LGTM

**Pull request merged**